include/veriblock/pop/blockchain/blockchain_util.hpp

```diff
@@ -111,13 +111,24 @@ bool recoverEndorsements(ProtectedBlockTree& ed_,
       auto& by = endorsed->getEndorsedBy();
       VBK_ASSERT_MSG(std::find(by.begin(), by.end(), endorsement) == by.end(),
                      "same endorsement is added to endorsedBy second time");
+      bool isDirty = endorsed->isDirty();
       endorsed->insertEndorsedBy(endorsement);
+      // keep dirty flag since recoverEndorsements is used when loading blocks from storage
+      // and should not affect dirtyness of the blocks
+      if (!isDirty) {
+        endorsed->unsetDirty();
+      }
 
       const auto& bop = blockOfProof->getBlockOfProofEndorsement();
       VBK_ASSERT_MSG(
           std::find(bop.begin(), bop.end(), endorsement) == bop.end(),
           "same endorsement is added to blockOfProof second time");
+      isDirty = blockOfProof->isDirty();
       blockOfProof->insertBlockOfProofEndorsement(endorsement);
+      // keep dirty flag
+      if (!isDirty) {
+        blockOfProof->unsetDirty();
+      }
```

Comment:
wait. it does't make sense.

if block was dirty,
 then unsetDirty won't be called.

if block was clean,
then unsetDirty will be called. 

Is this intentional?

Reply:
Yes it is intentional. We do not want recoverEndorsements to modify dirtyness of the block. So if it was dirty, it will remain dirty. And if it was clean, it will remain clean.